### DISCUSSION TITLE
offset 번호의 중복 / auto.offset.reset

### DISCUSSION THREAD

**Discussion by sejongjeong** [kyungjunleeme/kafka_study#24] [Q&A]
Kafka Offset 번호는 중복이 가능한가?

동일한 토픽내에서 partition 별로 offset을 관리하기 때문에 동일 토픽 동일 오프셋 존재는 당연하다(offset은 토픽 단위가 아니라 파티션 단위로 관리되는 정보)

**Comment by psh5487** [comment]
offset 번호 자체가 중복될 것 같지는 않은데, 
auto.offset.reset 값에 earlist 같은거를 주면, 배포 등으로 컨슈머가 붙었을 때 쌓여있는 메시지 중 맨 앞에 것을 읽어오면서, 중복으로 읽어올 위험이 있을 수 있겠네요.
그래서 auto.offset.reset 값에 latest 설정을 해주고 있습니다~!

**Comment by kyungjunleeme** [comment]
@psh5487 소현님 

> auto.offset.reset 값에 earlist 같은거를 주면, 배포 등으로 컨슈머가 붙었을 때 쌓여있는 메시지 중 맨 앞에 것을 읽어오면서, 중복으로 읽어올 위험이 있을 수 있겠네요.
그래서 auto.offset.reset 값에 latest 설정을 해주고 있습니다~!

해당 부분은 아래와 같은 이유를 근거로 메시지 유실을 위해서 `earliest`로 설정 해야지 메시지 유실을 막는다고 알고 있어서 저는 이렇게 설정을 해두었는데요. 혹시 `earliest`로 설정해서 메시지 중복을 경험하신 적이 있어서, `latest`설정을 하셨을까요?

https://blog.voidmainvoid.net/514

![image](https://github.com/user-attachments/assets/742b6a78-2d5c-4130-b68a-282e1692d834)
https://kafka.apache.org/documentation/#consumerconfigs_auto.offset.reset

**Comment by kyungjunleeme** [comment]
@hellices 시간 괜찮으실때 인환님도 의견 부탁드리겠습니다. (꾸벅)

**Reply by psh5487** [reply]
earlist로 하면 컨슈머가 토픽에 다시 붙을 때 토픽에 쌓여있는 것들을 처음부터 읽다보니, 이미 처리된 메시지를 또 읽어올 수 있어요.
하지만 유실이 절대 발생하면 안되는 상황이라면, earlist로 하고 같은 메시지 또 듣도록 하고, 컨슈머에서 이미 처리된 메시지라면 skip 되도록 방어 코드 써두는 방법도 있겠네여!

**Comment by hellices** [comment]
메시지 유실의 관점에서 위 말씀하신 내용으로 하는 경우, earlist는 처음부터 읽겠으나 kafka에서 메시지 보존 기간을 고려해야하는 점이 있을 것 같습니다. 며칠이 지나면 아무래도 데이터가 유실될 가능성이 높아지겠네요. 그래서 저는 토픽에 들어간 데이터는 최대한 빨리 끄집어내도록 처리해야한다고 봅니다.

저희의 경우에는 kafka를 cdc 위주로 사용하고 있는데 이러한 경우에는 redo log로 처리하는 로직으로 일부 겹치는 offset등에 대해서는 타깃 서버에서 pk 등으로 오류가 발생하기 때문에 offset보다는 순차 실행에 더 중요도를 높게 생각하며, 처리가 잘못된 경우에는 consumer에서 처리된 이력을 모두 삭제(truncate)해서 다시 수행하도록 하고 있습니다.
